include/swift/Runtime/Mutex.h

```diff
@@ -0,0 +1,313 @@
+//===--- Mutex.h - Lockables ------------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Mutex, Condition, and Scoped lock abstactions for use in Swift runtime.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_RUNTIME_MUTEX_H
+#define SWIFT_RUNTIME_MUTEX_H
+
+namespace swift {
+
+/// A Condition that works with Mutex to allow – as an example – multi-threaded
+/// producers and consumers to signal each other in a safe way.
+class Condition {
+  friend class MutexImplementation;
+
+public:
+  explicit Condition();
+  ~Condition();
+
+public:
+  /// Notifies one waiter (if any exists) that the condition has been met.
+  ///
+  /// Note: To avoid missed notification it is best hold the related mutex
+  //        lock when calling notifyOne.
+  void notifyOne() const;
+
+  /// Notifies all waiters (if any exists) that the condition has been met.
+  ///
+  /// Note: To avoid missed notification it is best hold the related mutex
+  //        lock when calling notifyAll.
+  void notifyAll() const;
+
+protected:
+  void *Data;
+};
+
+/// Internal (private) implementation of mutex functionality.
+/// Use Mutex or RecursiveMutex instead (see below).
```

Comment:
This comment includes a reference to the RecursiveMutex type that is no longer part of this change.